docs: fixup the builtin tutorial (#7488)
tendermint/tendermint: docs/tutorials/go-built-in.md

Core application from scratch. It does not assume that you have any prior
experience with Tendermint Core.

Tendermint Core is a service that provides a Byzantine fault tolerant consensus engine
for state-machine replication. The replicated state-machine, or "application", can be written 

> in any language that can send and receive protocol buffer messages. 

Reviewer: ```suggestion
in any language that can send and receive protocol buffer messages.
```

This tutorial is written for Go and uses Tendermint as a library, but applications not
written in Go can use Tendermint to drive state-machine replication in a client-server
model.

This tutorial expects some understanding of the Go programming language. 
If you have never written Go, you may want to go through [Learn X in Y minutes
Where X=Go](https://learnxinyminutes.com/docs/go/) first to familiarize
yourself with the syntax.

By following along with this guide, you'll create a Tendermint Core application
called kvstore, a (very) simple distributed BFT key-value store.

> > Note: please use a released version of Tendermint with this guide. The guides will work with the latest releasetd version.

Reviewer: ```suggestion
> Note: please use a released version of Tendermint with this guide. The guides will work with the latest released version.
```

> > Please, do not use master.

Reviewer: Or at least:
```suggestion
> Be aware that they may not apply to unreleased changes on master.
> We strongly advise against using unreleased commits for your development.
```

Core application from scratch. It does not assume that you have any prior
experience with Tendermint Core.

Tendermint Core is a service that provides a Byzantine fault tolerant consensus engine
for state-machine replication. The replicated state-machine, or "application", can be written 
in any language that can send and receive protocol buffer messages. 

> This tutorial is written for Golang and uses Tendermint as a library, but applications not

Reviewer: ```suggestion
This tutorial is written for Go and uses Tendermint as a library, but applications not
```

```bash
$ go version

> go version go1.17.x darwin/amd64

Reviewer: We should probably give the "real" output, even if it may be out of date.
```suggestion
go version go1.17.5 darwin/amd64
```

```

> 

Reviewer: ```suggestion

Note that the exact patch number may differ as Go releases come out.
```

## 1.3 Writing a Tendermint Core application

Tendermint Core communicates with an application through the Application

> BlockChain Interface (ABCI). All of the message types Tendermint uses for 

Reviewer: ```suggestion
BlockChain Interface (ABCI) protocol. All of the message types Tendermint uses for 
```

communicating with the application can be found in the ABCI [protobuf
file](https://github.com/tendermint/spec/blob/b695d30aae69933bc0e630da14949207d18ae02c/proto/tendermint/abci/types.proto).

We will begin by creating the basic scaffolding for an ABCI application in 
a new `app.go` file. The first step is to create a new type, `KVStoreApplication`

> with the methods that implement the abci `Application` interface.

Reviewer: ```suggestion
with methods that implement the abci `Application` interface.
```

 return abcitypes.ResponseInitChain{}
}

func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
 return abcitypes.ResponseBeginBlock{}
}

func (app *KVStoreApplication) EndBlock(req abcitypes.RequestEndBlock) abcitypes.ResponseEndBlock {
 return abcitypes.ResponseEndBlock{}
}

func (app *KVStoreApplication) ListSnapshots(abcitypes.RequestListSnapshots) abcitypes.ResponseListSnapshots {
 return abcitypes.ResponseListSnapshots{}
}

func (app *KVStoreApplication) OfferSnapshot(abcitypes.RequestOfferSnapshot) abcitypes.ResponseOfferSnapshot {
 return abcitypes.ResponseOfferSnapshot{}
}

func (app *KVStoreApplication) LoadSnapshotChunk(abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
 return abcitypes.ResponseLoadSnapshotChunk{}
}

func (app *KVStoreApplication) ApplySnapshotChunk(abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
 return abcitypes.ResponseApplySnapshotChunk{}
}
```

### 1.3.1 Add a persistent data store

Our application will need to write its state out to persistent storage so that it
can stop and start without losing all of its data.

> For this tutorial, we will use [badger](https://github.com/dgraph-io/badger). 

Reviewer: ```suggestion
For this tutorial, we will use [BadgerDB](https://github.com/dgraph-io/badger). 
```

Badger is a fast embedded key-value store. 

First, add badger as a dependency of your go module using the `go get` command:

> `go get github.com/dgraph-io/badger`

Reviewer: The current release is `/v3`. The API changes aren't too substantial, so it might be worth doing that update along with the other changes here. But YMMV.

Next, let's update the application and its constructor to receive a handle to the 
database. 

Update the application struct as follows: 

```go
type KVStoreApplication struct {
	db           *badger.DB

> 	currentBatch *badger.Txn

Reviewer: Would `currentBlock` be more accurate in this case? It is, as you note, a batch of transactions, but from the app's perspective that's sort of an implementation detail.

Maybe also worth a comment here:

```suggestion
	currentBatch *badger.Txn  // accumulates the transactions in the current block
```

}
```

And change the constructor to set the appropriate field when creating the application:

```go
func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {

> 	return &KVStoreApplication{
> 		db: db,
> 	}

Reviewer: ```suggestion
	return &KVStoreApplication{db: db}
```

}
```

> Don't worry about `currentBatch` for now, we'll get to that later.

Reviewer: It's probably still worth saying what this is, at a high level, even if we don't go into detail yet. Perhaps:

> The `currentBatch` keeps track of the transactions that update the application's state when a block is completed. Don't worry about it for now, we'll go into more detail later.

### 1.3.1 CheckTx

When Tendermint Core receives a new transaction, Tendermint asks the application
if the transaction is acceptable. In our new application, let's implement some

> basic validation for the transactions it will receive.

Reviewer: Maybe say something about the structure of transactions?

E.g.,
> For our KV store application, a transaction is a string like `key=value` giving a key and value to write to the store.

Add the following helper method to `app.go`:

```go
func (app *KVStoreApplication) validateTx(tx []byte) uint32 {
	parts := bytes.Split(tx, []byte("="))

	// first, check that the transaction is not malformed.
	if len(parts) != 2 {
		return 1
	}

	key, value := parts[0], parts[1]

	var txAlreadyExists bool

> 	// ok, we have a well-formed transaction, next let's check that we haven't

Reviewer: Hmm. I'm not sure we really want to do this. Shouldn't it be possible to update a key that was previously written? (Maybe not within the same block, but that's not really what we're checking here—this scans the already-stored data, not the pending dbtx).

	// already stored the same transaction in the database before.
	dbErr := app.db.View(func(txn *badger.Txn) error {
		item, err := txn.Get(key)
		if err != nil {
			if err != badger.ErrKeyNotFound {
				return err
			}
			return nil
		}
		return item.Value(func(val []byte) error {
			if bytes.Equal(val, value) {
				txAlreadyExists = true
			}
			return nil
		})
	})

	// There was an error processing the transaction.
	// The application therefore cannot safely make progress so we crash.
	if dbErr != nil {

> 		panic(dbErr)

Reviewer: Maybe we can demonstrate better error messages?
```suggestion
		panic(fmt.Sprintf("Error reading database, unable to verify tx: %v", dbErr))
```

Or maybe `log.Panicf`, for applications not stuck on zerolog.

	}

	if txAlreadyExists {
		return 2
	}
	return 0

}
```

And call it from within your `CheckTx` method:

```go
func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
	if code := app.validateTx(req.Tx); code != 0 {
		return abcitypes.ResponseCheckTx{Code: code}
	}
	return abcitypes.ResponseCheckTx{Code: 0}
}
```

> Any response with a non-zero code will be considered invalid by Tendermint. 

Reviewer: Are these error codes standardized at all, or is it just between the application and its own bad self? (I should probably already know the answer to this)

> Our `CheckTx` logic returns `0` to Tendermint when the transaction passes
> the validation checks.

Reviewer: ```suggestion
Our `CheckTx` logic returns `0` to Tendermint when a transaction passes
its validation checks.
```

It may also be worth reinforcing:

> Note that `CheckTx` does not _execute_ the transaction, it only verifies
> that the transaction _could_ be executed.  We don't know yet, at this point,
> whether the rest of the network has agreed to accept this transaction into
> a block.

will be delivered to the application.

To implement these calls in our application we're going to make use of Badger's 
transaction mechanism. 

First, let's create a new Badger `Txn` during `BeginBlock`:

```go
func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
	app.currentBatch = app.db.NewTransaction(true)
	return abcitypes.ResponseBeginBlock{}
}
```

Next, let's modify `DeliverTx` to add the `key` and `value` to this `Txn` every time our application
receives a new `RequestDeliverTx`.

```go
func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
	if code := app.validateTx(req.Tx); code != 0 {
		return abcitypes.ResponseDeliverTx{Code: code}
	}

	parts := bytes.Split(req.Tx, []byte("="))
	key, value := parts[0], parts[1]

	err := app.currentBatch.Set(key, value)
	if err != nil {
		panic(err)
	}

	return abcitypes.ResponseDeliverTx{Code: 0}
}
```
Note that we check the validity of the transaction _again_ during `DeliverTx`.
Transactions are not guaranteed to be valid when they are delivered to to an

> application. 

Reviewer: Is there anything we can say here about why that could be, even if only cursory?

I don't think we need to give deep details, but having a general sense of how a tx could be invalid at delivery even if it passed `CheckTx` might help build intuition.

Also note that we don't commit the Badger `Txn` we are building during `DeliverTx`. 
Other methods, such as `Query`, rely on a consistent view of the application's state.
The application should only update its state when the full block has been delivered.

The `Commit` method indicates that the full block has been delivered. During `Commit`,
the application should persist the pending `Txn`.

Let's modify our `Commit` method to persist the new state to the database:

```go
func (app *KVStoreApplication) Commit() abcitypes.ResponseCommit {

> 	app.currentBatch.Commit()

Reviewer: This can fail, and we should check.

In particular: It is possible the dbtx has conflicts.

	return abcitypes.ResponseCommit{Data: []byte{}}
}
```

### 1.3.3 Query Method

We'll want to be able to determine if a transaction was committed to the state-machine.
To do this, let's implement the `Query` method in `app.go`:

```go
func (app *KVStoreApplication) Query(req abcitypes.RequestQuery) abcitypes.ResponseQuery {

> 	var resp abcitypes.ResponseQuery
> 	resp.Key = req.Data

Reviewer: ```suggestion
	resp := abcitypes.ResponseQuery{Key: req.Data}
```

	dbErr := app.db.View(func(txn *badger.Txn) error {
		item, err := txn.Get(req.Data)
		if err != nil {
			if err != badger.ErrKeyNotFound {
				return err
			}
			resp.Log = "key does not exist"
			return nil
		}

		return item.Value(func(val []byte) error {
			resp.Log = "exists"
			resp.Value = val
			return nil
		})
	})
	if dbErr != nil {
		panic(dbErr)
	}
	return resp
}
```

## 1.3.4 Additional Methods

You'll noticed that we left several methods unchanged. Specifically, we have yet
to implement the `Info` and `InitChain` methods and we did not implement 
any of the `*Snapthot` methods. These methods are all important for running Tendermint
applications in production but are not required for getting a very simple application
up and running. 

To better understand these methods and why they are useful, check out the Tendermint 

> [specification on ABCI]()

Reviewer: Missing link?

Author: yup :(

## 1.4 Starting an application and a Tendermint Core instance in the same process

Now that we have the basic functionality of our application in place, let's put it
all together inside of our `main.go` file.

Add the following code to your `main.go` file: 

```go
package main

import (
	"flag"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/dgraph-io/badger"
	"github.com/spf13/viper"
	abciclient "github.com/tendermint/tendermint/abci/client"
	cfg "github.com/tendermint/tendermint/config"
	"github.com/tendermint/tendermint/libs/log"
	nm "github.com/tendermint/tendermint/node"
	"github.com/tendermint/tendermint/types"
)

var homeDir string

func init() {
	flag.StringVar(&homeDir, "tm-home", "", "Path to the tendermint 'home' directory")
}

func main() {
	flag.Parse()
	if homeDir == "" {
		h, err := os.UserHomeDir()
		if err != nil {
			panic(err)
		}
		homeDir = fmt.Sprintf("%s/%s", h, ".tendermint")
	}
	config := cfg.DefaultValidatorConfig()

	config.SetRoot(homeDir)

	viper.SetConfigFile(fmt.Sprintf("%s/%s", homeDir, "config/config.toml"))
	if err := viper.ReadInConfig(); err != nil {

> 		panic(err)

Reviewer: These panics should probably all be `log.Fatal` instead; a goroutine dump isn't valuable in debugging "expected" infra failures.

```suggestion
		log.Fatalf("Reading config: %v", err)
```

	}
	if err := viper.Unmarshal(config); err != nil {

> 		panic(err)

Reviewer: ```suggestion
		log.Fatalf("Decoding config: %v", err)
```

	}
	if err := config.ValidateBasic(); err != nil {

> 		panic(err)

Reviewer: ```suggestion
		log.Fatalf("Invalid configuration data: %v", err)
```

	}
	gf, err := types.GenesisDocFromFile(config.GenesisFile())
	if err != nil {

> 		panic(err)

Reviewer: ```suggestion
		log.Fatalf("Loading genesis document: %v", err)
```

	}

> 	db, err := badger.Open(badger.DefaultOptions("./badger").WithTruncate(true))

Reviewer: Shouldn't this be inside `tm-home`?
```suggestion
	dbPath := filepath.Join(homeDir, "badger")
	db, err := badger.Open(badger.DefaultOptions(dbPath).WithTruncate(true))
```

	if err != nil {

> 		panic(err)

Reviewer: ```suggestion
		log.Fatalf("Opening database: %v", err)
```

	}

> 	defer db.Close()

Reviewer: This error should be checked. Per [the docs](https://pkg.go.dev/github.com/dgraph-io/badger/v3#DB.Close).

> It's crucial to call it to ensure all the pending updates make their way to disk. 

…and the operation definitely can meaningfully fail (I checked the code).

Author: fixed

go build -mod=mod -o my-app
```

Everything is now in place to run your application.

Run: 
```bash
$ ./my-app -tm-home ./tendermint-home
```

The application will begin producing blocks and you can see this reflected in
the log output.

You now have successfully started running an application using Tendermint Core 🎉🎉.

## 1.6 Using the application

Your application is now running and emitting logs to the terminal.
Now it's time to see what this application can do!

Let's try submitting a transaction to our new application.

Open another terminal window and run the following curl command:

```bash
$ curl -s 'localhost:26657/broadcast_tx_commit?tx="tendermint=rocks"'
```

If everything went well, you should see a response indicating which height the 
transaction was included in the blockchain.

Finally, let's make sure that transaction really was persisted by the application.

Run the following command:

```bash
$ curl -s 'localhost:26657/abci_query?data="tendermint"'
```

Let's examine the response object that this request returns.
The request returns a `json` object with a `key` and `value` field set.

```json
...
      "key": "dGVuZGVybWludA==",
      "value": "cm9ja3M=",

>  ...

Reviewer: ```suggestion
...
```

Next, let's update the application and its constructor to receive a handle to the 
database. 

Update the application struct as follows: 

```go
type KVStoreApplication struct {
	db           *badger.DB
	pendingBlock *badger.Txn
}
```

And change the constructor to set the appropriate field when creating the application:

```go
func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
	return &KVStoreApplication{db: db}
}
```

The `pendingBlock` keeps track of the transactions that will update the application's
state when a block is completed. Don't worry about it for now, we'll get to that later.

### 1.3.1 CheckTx

When Tendermint Core receives a new transaction, Tendermint asks the application
if the transaction is acceptable. In our new application, let's implement some
basic validation for the transactions it will receive.

For our KV store application, a transaction is a string with the form `key=value`,
indicating a key and value to write to the store.

Add the following helper method to `app.go`:

```go
func (app *KVStoreApplication) validateTx(tx []byte) uint32 {
	parts := bytes.SplitN(tx, []byte("="), 2)

	// check that the transaction is not malformed

> 	if len(parts) != 2 {

Reviewer: Possibly also:
```suggestion
	if len(parts) != 2 || len(parts[0]) == 0 {
```

to forestall an empty key.

	code := app.validateTx(req.Tx)
	return abcitypes.ResponseCheckTx{Code: code}
}
```

Any response with a non-zero code will be considered invalid by Tendermint.
Our `CheckTx` logic returns `0` to Tendermint when a transaction passes
its validation checks. The specific value of the code is meaningless to Tendermint.
Non-zero codes are logged by Tendermint so applications can provide more specific
information on why the transaction was rejected.

Note that `CheckTx` _does not execute_ the transaction, it only verifies that that the
transaction _could_ be executed. We do not know yet if the rest of the network has
agreed to accept this transaction into a block. 

Finally, make sure to add the `bytes` package to the your import stanza
at the top of `app.go`:

```go
import (
 "bytes"

 abcitypes "github.com/tendermint/tendermint/abci/types"
)
```

While this `CheckTx` is simple and only validates that the transaction is well-formed,
it is very common for `CheckTx` to make more complex use of the state of an application.

### 1.3.2 BeginBlock -> DeliverTx -> EndBlock -> Commit

When the Tendermint consensus engine has decided on the block, the block is transferred to the

> application over three ABCI method calls: `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer: Probably something for a future iteration, but it occurs to me this is probably where we should be checking for duplicate keys. Otherwise, we accept a Tx only to fail when the block is committed, and the chain stalls.

Author: Not totally sure I understand the issue being described here?

Reviewer: Actually I think this is no longer an issue. I was concerned because we were previously failing for txn that updated an existing key. Now that check is gone, it's probably not important to worry about here.

The corner case is smaller now: A database tx may get conflicts if we try to include multiple txn for the same key _in a single block_, but for demonstration that's not super critical.

```go
...
	config := cfg.DefaultValidatorConfig()

	config.SetRoot(homeDir)

	viper.SetConfigFile(fmt.Sprintf("%s/%s", homeDir, "config/config.toml"))
	if err := viper.ReadInConfig(); err != nil {
		log.Fatalf("Reading config: %v", err)
	}
	if err := viper.Unmarshal(config); err != nil {
		log.Fatalf("Decoding config: %v", err)
	}
	if err := config.ValidateBasic(); err != nil {
		log.Fatalf("Invalid configuration data: %v", err)
	}
	gf, err := types.GenesisDocFromFile(config.GenesisFile())
	if err != nil {
		log.Fatalf("Loading genesis document: %v", err)
	}
...
```

Next, we create a database handle and use it to construct our ABCI application:

```go
...
	dbPath := filepath.Join(homeDir, "badger")
	db, err := badger.Open(badger.DefaultOptions(dbPath).WithTruncate(true))
	if err != nil {
		log.Fatalf("Opening database: %v", err)
	}
	defer func() {
		if err := db.Close(); err != nil {
			log.Fatalf("Error closing database: %v", err)
		}
	}()
	app := NewKVStoreApplication(db)
	acc := abciclient.NewLocalCreator(app)
...
```

Then we construct a logger:
```go
...
	logger := log.MustNewDefaultLogger(log.LogFormatPlain, log.LogLevelInfo, false)
...
```

> Now we have everything setup to run the Tendermint Core node. We construct

Reviewer: ```suggestion
Now we have everything setup to run the Tendermint node. We construct
```

a node by passing it the configuration, the logger, a handle to our application and
the genesis file:

```go
...
	node, err := nm.New(config, logger, acc, gf)
	if err != nil {
		log.Fatalf("Creating node: %v", err)
	}
...
```

Finally, we start the node:
```go
...
	node.Start()
	defer func() {
		node.Stop()
		node.Wait()
	}()
...
```

The additional logic at the end of the file allows the program to catch `SIGTERM`.
This means that the node can shutdown gracefully when an operator tries to kill the program:

```go 
...
c := make(chan os.Signal, 1)
signal.Notify(c, os.Interrupt, syscall.SIGTERM)
<-c
...
```

## 1.5 Getting Up and Running

Our application is almost ready to run.

> Let's instruct the `go` tooling to use the correct version of the Tendermint library.

Reviewer: ```suggestion
Let's install the latest release version of the Tendermint library.
```

From inside of the project directory, run:

```sh
go get github.com/tendermint/tendermint@v0.35.0
```

Next, we'll need to populate the Tendermint Core configuration files.
This command will create a `tendermint-home` directory in your project and add a basic set of configuration
files in `tendermint-home/config/`. For more information on what these files contain
see [the configuration documentation]()

From the root of your project, run:
```bash
go run github.com/tendermint/tendermint/cmd/tendermint@v0.35.0 init validator --home ./tendermint-home
```

Next, build the application: 
```bash

> go build -mod=mod -o my-app

Reviewer: ```suggestion
go build -mod=mod -o my-app  # use -mod=mod to automatically update go.sum
```
